extensions/parquet/compression/src/main/java/io/deephaven/parquet/compress/LZ4WithLZ4RawBackupCompressorAdapter.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.parquet.compress;
+
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.parquet.bytes.BytesInput;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * This is the default adapter for LZ4 files. It attempts to decompress with LZ4 and falls back to LZ4_RAW on failure.
+ * After that, it always uses LZ4_RAW for decompression. This fallback mechanism is particularly useful for
+ * decompressing parquet files that are compressed with LZ4_RAW but tagged as LZ4 in the metadata. This adapter is
+ * internally stateful in some cases and therefore a single instance should not be re-used across files.
+ */
+class LZ4WithLZ4RawBackupCompressorAdapter extends DeephavenCompressorAdapterFactory.CodecWrappingCompressorAdapter {
+    private enum DecompressionMode {
+        INIT, LZ4, LZ4_RAW
+    }
+
+    private DecompressionMode mode = DecompressionMode.INIT;
+
+    /**
+     * Only initialized if we hit an exception while decompressing with LZ4.
+     */
+    private CompressorAdapter lz4RawAdapter = null;
+
+    LZ4WithLZ4RawBackupCompressorAdapter(CompressionCodec compressionCodec,
+            CompressionCodecName compressionCodecName) {
+        super(compressionCodec, compressionCodecName);
+    }
+
+    @Override
+    public BytesInput decompress(final InputStream inputStream, final int compressedSize,
+            final int uncompressedSize) throws IOException {
+        if (mode == DecompressionMode.LZ4) {
+            return super.decompress(inputStream, compressedSize, uncompressedSize);
+        }
+        if (mode == DecompressionMode.LZ4_RAW) {
+            assert lz4RawAdapter != null;
```

Comment:
I did not want to add a new gradle dependency on `Deephaven_Community_Core.Base.main`, that is why using this `assert` instead of `Assert.neqNull`.
Let me know if I should add that dependency.

Reply:
Should probably throw IllegalStateException/IllegalArgumentException then - `assert` is turned off by default at runtime, so this will not throw in production.